### PR TITLE
Add third party content support

### PR DESCRIPTION
**Pull request by semtexzv**

### CODE REVIEW

**Comment by psegedy** [comment]
breaks errata endpoint
```
POST /api/v3/errata, body={"errata_list": ["RHSA-2017:1931"]}

Response: {'type': 'about:blank', 'detail': 'Internal server error <-9223363283135254769>: please include this error id in bug report.', 'status': 500}
```

**Review by Josca — CHANGES_REQUESTED** [review]

**Comment by jdobes** [comment]
pls rebase

File: webapp/webapp.v3.spec.yaml [hunk 1]
```diff
@@ -670,6 +674,10 @@ components:
           example: '2018-04-05T01:23:45+02:00'
           format: date-time
           description: Return only repositories changed after the given date
+        third_party:
+          type: boolean
+          enum: [ true, false ]
+          description: Return third party content (Content not managed by RedHat)
```
Comment (Josca) [inline webapp/webapp.v3.spec.yaml:680]: I would add the info that param is optional and false by default.

**Review by Josca — CHANGES_REQUESTED** [review]
Looks good, just some small things found.

**Review by Josca — COMMENTED** [review]
I synced epel repos:
~~~json
[
    {
        "products": {
            "Extra Packages for Enterprise Linux (EPEL)": {
                "content_sets": {
                    "epel-7": {
                        "name": "Extra Packages for Enterprise Linux 7",
                        "baseurl": "http://download.fedoraproject.org/pub/epel/$releasever/$basearch",
                        "basearch": [
                            "x86_64",
                            "ppc64le"
                        ],
                        "releasever": [
                            "7Server",
                            "7"
                        ],
                        "third_party": true
                    },
                    "epel-8": {
                        "name": "Extra Packages for Enterprise Linux 8",
                        "baseurl": "https://download.fedoraproject.org/pub/epel/$releasever/Everything/$basearch",
                        "basearch": [
                            "x86_64",
                            "ppc64le"
                        ],
                        "releasever": [
                            "8"
                        ],
                        "third_party": true
                    },
                    "epel-8-modular": {
                        "name": "Extra Packages for Enterprise Linux Modular 8",
                        "baseurl": "https://download.fedoraproject.org/pub/epel/$releasever/Modular/$basearch",
                        "basearch": [
                            "x86_64",
                            "ppc64le"
                        ],
                        "releasever": [
                            "8"
                        ],
                        "third_party": true
                    }
                }
            }
        }
    }
]
~~~
.. then checked repo table and some repos have empty "revision" column. Is it ok?
~~~
 id |                                url                                | content_set_id | basearch_id | releasever | eol |        revision        | certificate_id | third_party 
----+-------------------------------------------------------------------+----------------+-------------+------------+-----+------------------------+----------------+-------------
  3 | http://download.fedoraproject.org/pub/epel/7Server/x86_64/        |              1 |          20 | 7Server    | f   |                        |                | t
  4 | http://download.fedoraproject.org/pub/epel/7Server/ppc64le/       |              1 |          17 | 7Server    | f   |                        |                | t
  8 | https://download.fedoraproject.org/pub/epel/8/Everything/ppc64le/ |              2 |          17 | 8          | f   | 2021-03-24 02:49:01+00 |                | t
  7 | https://download.fedoraproject.org/pub/epel/8/Everything/x86_64/  |              2 |          20 | 8          | f   | 2021-03-24 02:49:02+00 |                | t
  6 | http://download.fedoraproject.org/pub/epel/7/ppc64le/             |              1 |          17 | 7          | f   | 2021-03-24 01:53:39+00 |                | t
  5 | http://download.fedoraproject.org/pub/epel/7/x86_64/              |              1 |          20 | 7          | f   | 2021-03-24 01:53:49+00 |                | t
  1 | https://download.fedoraproject.org/pub/epel/8/Modular/x86_64/     |              3 |          20 | 8          | f   | 2021-03-24 02:57:43+00 |                | t
  2 | https://download.fedoraproject.org/pub/epel/8/Modular/ppc64le/    |              3 |          17 | 8          | f   | 2021-03-24 02:57:43+00 |                | t
(8 rows)
~~~

File: webapp/webapp.v3.spec.yaml [hunk 2]
```diff
@@ -504,6 +504,10 @@ components:
         latest_only:
           type: boolean
           enum: [true, false]
+        third_party:
```
Comment (Josca) [inline webapp/webapp.v3.spec.yaml:507]: Does "third_party" flag in requests take any effect? I tried errata, cve, packages endpoints without third_party flag set and also set to false and for all I got data also from third party repositories what's wrong I guess.

**Review by Josca — CHANGES_REQUESTED** [review]
I tested current version. It seems to work properly but when epel data are synced I found inconsistent response with master:
/packages request:
~~~
{
    "package_list": [
      "ncdu-1.15.1-1.el7.x86_64"
    ],
    "third_party": false
}
Response: 
~~~
{"package_list": {}}
~~~
Expected response (master):
~~~
{"package_list": {"ncdu-1.15.1-1.el7.x86_64": {}}}
~~~
...  it should included requested NEVRAs into the response, but when they are third_party they are excluded.

File: webapp/webapp.v3.spec.yaml (hunk 1 above)
Comment (semtexzv) [reply webapp/webapp.v3.spec.yaml:681]: Added

File: webapp/webapp.v3.spec.yaml [hunk 3]
```diff
@@ -1019,6 +1038,11 @@ components:
         oval_only:
           type: boolean
           enum: [true, false]
+        third_party:
```
Comment (jdobes) [inline webapp/webapp.v3.spec.yaml:1041]: defined here but not used in vulnerabilities API

File: webapp/packages.py [hunk 4]
```diff
@@ -36,12 +37,17 @@ def process_list(self, api_version, data): # pylint: disable=unused-argument
         :returns: json response with package details
         """
         packages = data.get('package_list', None)
+        # By default, don't include third party data
+        want_third_party = data.get('third_party', False)
```
[the remark this answers is not in the document]
Comment (semtexzv) [reply webapp/packages.py:41]: Added

File: webapp/webapp.v3.spec.yaml (hunk 3 above)
Comment (semtexzv) [reply webapp/webapp.v3.spec.yaml:1041]: It is. Vulnerabilities internally uses the updates api, and that api supports third party flag.

File: database/vmaas_db_postgresql.sql [hunk 5]
```diff
@@ -527,6 +528,7 @@ CREATE TABLE IF NOT EXISTS errata (
   solution TEXT, CHECK (NOT empty(solution)),
   issued TIMESTAMP WITH TIME ZONE NOT NULL,
   updated TIMESTAMP WITH TIME ZONE NOT NULL,
+  third_party BOOLEAN NOT NULL DEFAULT FALSE,
```
[the remark this answers is not in the document]
Comment (semtexzv) [reply database/vmaas_db_postgresql.sql:531]: Package could potentially be in both RedHat and EPEL repositories. For advisories, they have different convnetions.

File: database/vmaas_db_postgresql.sql [hunk 6]
```diff
@@ -448,6 +448,7 @@ CREATE TABLE IF NOT EXISTS repo (
   eol BOOLEAN NOT NULL,
   revision TIMESTAMP WITH TIME ZONE NULL,
   certificate_id INT NULL,
+  third_party BOOLEAN NOT NULL DEFAULT FALSE,
```
[the remark this answers is not in the document]
Comment (semtexzv) [reply database/vmaas_db_postgresql.sql:451]: Maybe. It was much simpler to match it with errata when it was in repos.

File: webapp/repos.py [hunk 7]
```diff
@@ -48,6 +48,21 @@ def _filter_modified_since(self, repos_to_process, modified_since_dt):
...
+        filtered_repos = []
+
+        if third_party:
+            return repos_to_process
+
+        for label in repos_to_process:
+            for repo_id in self.cache.repolabel2ids.get(label, []):
+                repo_detail = self.cache.repo_detail[repo_id]
+                # If we don't want third party repo, and this repo is flagged as third party, skip it
+                if not third_party and repo_detail[REPO_THIRD_PARTY]:
```
Comment (Josca) [inline webapp/repos.py:61]: You don't need to check "third_party" again here. It's surely False.

File: webapp/repos.py [hunk 8]
```diff
@@ -48,6 +48,21 @@ def _filter_modified_since(self, repos_to_process, modified_since_dt):
                     break
         return filtered_repos_to_process
 
+    def _filter_third_party(self, repos_to_process, third_party):
+        filtered_repos = []
```
Comment (Josca) [inline webapp/repos.py:52]: It's better to define that after "if third_party" check.

File: webapp/test/data/cache.yml [hunk 9]
```diff
@@ -382,6 +390,7 @@ repo_detail:
   - null
   - null
   - '2019-09-20T06:25:12+00:00'
+  - false
```
Comment (Josca) [inline webapp/test/data/cache.yml:393]: All repos are third_party false here. Then it's hard to test whether new feature works properly. Could you add some third_party repo and update tests accordingly?

**Review by Josca — CHANGES_REQUESTED** [review]

File: webapp/repos.py (hunk 7 above)
Comment (semtexzv) [reply webapp/repos.py:61]: Well, maybe not, but it expresses the intent more clearly.

File: reposcan/reposcan.py [hunk 10]
```diff
@@ -216,22 +216,31 @@ class RepolistImportHandler(SyncHandler):
...
-        repos = [(baseurl, None, None)]
-        # Replace basearch
-        if basearches:
-            repos = [(repo[0].replace("$basearch", basearch), basearch, repo[2])
-                     for basearch in basearches for repo in repos]
-        # Replace releasever
-        if releasevers:
-            repos = [(repo[0].replace("$releasever", releasever), repo[1], releasever)
-                     for releasever in releasevers for repo in repos]
+        # Accept a list or single baseurl
```
[the remark this answers is not in the document]
Comment (semtexzv) [reply reposcan/reposcan.py:224]: We need multiple different baseurls for a repo with single identifier. EPEL for 7 and 8 has different basurls, but same label.

File: webapp/repos.py (hunk 7 above)
Comment (Josca) [reply webapp/repos.py:61]: I don't think so. It's just confusing.

File: reposcan/reposcan.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jdobes) [reply reposcan/reposcan.py:224]: @Josca we need it if we want single content set for all EPEL versions (on client side the repo is labeled as `epel` so I think this is needed)

but cannot use single baseurl, for example there is

https://dl.fedoraproject.org/pub/epel/7/x86_64/

but

https://dl.fedoraproject.org/pub/epel/8/Everything/x86_64/

File: webapp/test/data/cache.yml (hunk 9 above)
Comment (semtexzv) [reply webapp/test/data/cache.yml:409]: Added some test data for advisories and repos.

**Comment by semtexzv** [comment]
Splitting up to 2 PRs

**Pull request closed without merging**